### ISSUE TITLE
Vision: Redesign Event system

### ISSUE THREAD

**Issue by xlc** [paritytech/polkadot-sdk#278]
Related issues: https://github.com/paritytech/substrate/issues/9480 https://github.com/paritytech/substrate/issues/11132

Currently all the events are stored in a single array. This means while we can emit events in the runtime, we can never be able to process them in runtime or light client efficiently and safely. This prevents people from using foreign events during cross chain interactions and basically makes many async multi chain operations impossible.

We also have supports event topic that is incomplete (https://github.com/paritytech/substrate/issues/2553) and unused at all.

We should redesign the event system to meet at least those requirements:

- Efficient to emit
  - Emit event should have minimal overhead, should not increase PoV size.
- Efficient to proof
  - It should be possible to generate a proof of existent of some events with minimal proof size. Light client and runtime should be able to verify the proof very efficiently.
- Efficient removal
  - We need to purge all events from previous block at beginning of block construction so remove old events should be as efficient as possible.
- Easy to query
  - Clients and light clients should be able to query some particular events without iterating every single events in the block
  - Clients and light clients should be able to subscribe for some particular events by monitoring for storage changes for some storage keys
- O(1) for checkpointing/commit/reversion in transactional processing.

**Comment by kayabaNerve** [comment]
> RSA Accumulators almost meet all those requirements:

When I read this months ago (November?), I read up on RSA Accumulators. I was completely horrified. Not only do they have a trusted setup, the idea of hashing to prime numbers seemed like a mess. I did see there was groups of unknown order proposed as an alternative to a trusted setup, yet it was still quite fringe.

For completely unrelated reasons, I ended up working on threshold ECDSA and for that, implemented class groups of unknown order in Rust. With that, I implemented a hash to prime (and was again horrified). I ended up accepting it however for its role in removing cryptographic assumptions, having better understood them. Please note many academics seem to think that hash to prime isn't necessary for class groups, and do rely on the Strong Root Assumption (not just Adaptive Root).

If anyone is interested in using my work, which is in pure Rust, for experiments: https://github.com/kayabaNerve/threshold-threshold-ecdsa/blob/develop/src/class_group.rs

Please note I doubt my implementation is optimal. I believe Chia has made far more efficient reduction algorithms, and I know malachite is ~4x slower than gmp (though this would be a C dependency).

Please note hashing to prime numbers is still messy to implement, and I don't doubt it can be adversarially bombed. I do still question what the worst case would be for an algorithm designed to be used as a hash to prime is. Please note the proposed brute force above can be replaced with a sieve of some form, such as https://eprint.iacr.org/2011/481 which intends to effect a uniform distribution. 

Such a proposal could use a single storage slot, and create an inclusion proof with a trivial algorithm using the accumulator of the prior block + an O(n) algorithm where all other events in the same block are included (leaving out the intended event).

I'll also note there are trustless mechanisms to generate uniform elements in a class group, per https://eprint.iacr.org/2024/034.

---

The above is solely to flesh out the cryptographic proposal of accumulators. My commentary on cuckoo filters is it doesn't sound to actually offer LC verification, solely efficient finding of where events may have been. Is that correct, or would the light client proof be all of the hashes in the filter (where the event's hash inclusion proved it occurred)?

**Comment by NingLin-P** [comment]
We also hit this issue with `Events` being a big single value in the trie. In our case, we need to calculate and collect the storage root after executing each extrinsic (as an execution trace for potential fraud proof usage), and calculating the storage root gets slower and slower as more extrinsic are executed and the `Events` array gets bigger and bigger.

I saw this issue has been discussed multiple times and the current trend seems to be moving the event offchain with [this feature](https://github.com/paritytech/polkadot-sdk/issues/245), but before it lands (which may take quite some time) I wonder if we can approach with a simpler solution/workaround that can partially resolve the issue.

For example, replacing `StorageValue<Vec<Event>>` with `CountedStorageMap<u32, Event>` so each event is present as a single value in the trie, which can archive some of the goals of this Vision, see [this PR](https://github.com/autonomys/polkadot-sdk/pull/22) for the changes.

**Comment by bkchr** [comment]
> For example, replacing `StorageValue<Vec<Event>>` with `CountedStorageMap<u32, Event>` so each event is present as a single value in the trie, which can archive some of the goals of this Vision, see [this PR](https://github.com/autonomys/polkadot-sdk/pull/22) for the changes.

The problem with this is that the removal of the old events would be extremely costly as it requires an iteration over the storage to find all events. Hmm okay, while writing this, I had an idea. We could count the number of events in an extra storage item and then just delete the keys without a storage iteration. However, this should be benchmarked to see what the impact of this is.

**Comment by NingLin-P** [comment]
>We could count the number of events in an extra storage item and then just delete the keys without a storage iteration.

We already have that, the purpose of using `CountedStorageMap` is to use `0..count` as the key of the map, which also helps preserve the order of the event (in case it matters somehow).

>However, this should be benchmarked to see what the impact of this is.

I would love to help with this!

**Comment by bkchr** [comment]
> using `CountedStorageMap` is to use `0..count`

I missed `CountedStorageMap` and just read it as `StorageMap` 🙈 



> I would love to help with this!

Yeah feel free to write some benchmarks for this.

**Comment by NingLin-P** [comment]
Hey, sorry for the delay! I have done some benchmarks:

By simply using `CountedStorageMap` to store each event under a single key, it seems to bring performance regression to both block execution and block initialization (for clearing the event). The regression in block execution is unexpected and my guess is there are too many events emitted (e.g. >50k events after 8k of transfer) and as they each occupied one key in the state, it made the overlay change much larger and operation getting slower and slower.

I found a way to mitigate regression by segmenting the events to store more than one event under a single key (but still much less than storing all events under a single key), 100 events per key seems to be a good parameter to eliminate regression on block execution and also improve performance of block initialization.

The following are the benchmark results (for a block with 8k transfers running on a `MacBook Pro, Apple M1 Pro, 10 Core(s)`):

 \ | Current Events | Events map | Events segmentation 
-- | -- | -- | -- 
Block execution | 2_030ms | 2_355ms | 2_034ms 
Block initialization | 1ms | 174ms | 112ms 

You can reproduce by pulling [this branch](https://github.com/NingLin-P/polkadot-sdk/tree/refactoring-events) and for the last 3 commits (which represent the `Current Events`, `Events map`, and `Events segmentation` respectively):
```
// Build the template node
cargo build -r --bin substrate-node

// Run the benchmark to execute a block with 8000 txs and its next block
./target/release/substrate-node benchmark-block-execution --number 8000 --dev
```

You can also use `cargo bench` under [this crate](https://github.com/paritytech/polkadot-sdk/tree/master/substrate/bin/node/cli) to bench block execution and block initialization (for clearing events):
```
cargo bench --bench block_production
cargo bench --bench block_initialization
```

**Comment by xlc** [comment]
maybe we don't need to actually delete events in init? we just set the count to zero.

**Comment by NingLin-P** [comment]
>maybe we don't need to actually delete events in init? we just set the count to zero.

Sounds like a good idea!

Since we have an `EventCount` to keep track of events emitted in the current block we won't accidentally access events emitted from the previous blocks. And by introducing another tracker `UnclearedEventCount` we can know there are how many uncleared events from the previous block and we can periodically clear these events or introduce an extrinsic to clear the event manually.

**Comment by NingLin-P** [comment]
Ping @bkchr can I have your feedback on this, please?

**Comment by bkchr** [comment]
> Sounds like a good idea!

Did you also benchmarked this?

**Comment by NingLin-P** [comment]
Just benchmarked, see the `Events segmentation and not clearing events` in https://github.com/paritytech/polkadot-sdk/issues/278#issuecomment-2665929051, also see the change [here](https://github.com/NingLin-P/polkadot-sdk/commit/498d13637e95806ab8c97949d92e41edebcadf82)

**Comment by bkchr** [comment]
I can not think of any real downside of switching to the proposed solution. Yes, that we maybe keep events around is not such a great point, but I also don't have any better idea for this right now.

One thing we should consider is that we we use the `IdentityHash` for the event number. This will make events ordered in the trie and if you need to proof that event X happened for extrinsic Y better, because they should both be near to each other in the tree.

@NingLin-P I assume you would implement this?

**Comment by bkchr** [comment]
One big downside is the impact of this for anyone reading events offchain right now, aka basically anyone.

**Comment by NingLin-P** [comment]
>One big downside is the impact of this for anyone reading events offchain right now, aka basically anyone.

Yeah, that is unfortunate but also inevitable as the storage structure has to change.

Will open a PR for all these changes.

**Comment by xlc** [comment]
we need feature flag to toggle this to avoid forcing chains to optin this major breaking change

**Comment by bkchr** [comment]
> we need feature flag to toggle this to avoid forcing chains to optin this major breaking change

We could probably add some option to the `Config` trait that changes how the events are being stored.
